### PR TITLE
WT-10961 Fix OOO keys caused by racing deletion and insertion on left subtrees

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

**Comment by quchenhao** [comment]
https://spruce.mongodb.com/version/645c3da47742ae3793c79c13/tasks?page=0&sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC&statuses=success&variant=%5Eubuntu2004-stress-tests%24 stress test

File: src/include/btree_inline.h [hunk 1]
```diff
@@ -2158,7 +2158,7 @@ __wt_split_descent_race(WT_SESSION_IMPL *session, WT_REF *ref, WT_PAGE_INDEX *sa
      * this code, don't re-order that acquisition with this check.
      */
     WT_BARRIER();
-    WT_INTL_INDEX_GET(session, ref->home, pindex);
+    pindex = WT_INTL_INDEX_GET_SAFE(ref->home);
```
Comment (Siddhartha8899) [inline src/include/btree_inline.h:2161]: Could you please give the reason for the change in the macro? It seems like you wanted to exclude the following assertion.
`  WT_ASSERT(session, __wt_session_gen(session, WT_GEN_SPLIT) != 0);`

Comment (quchenhao) [reply src/include/btree_inline.h:2161]: for this call
```
if (__wt_split_descent_race(cbt->ref, pindex))
            skiplow = skiphigh = 0;
```
we have descended to the leaf page and has exited the split gen.

File: src/btree/row_srch.c [hunk 2]
```diff
@@ -389,9 +390,9 @@ __wt_row_search(WT_CURSOR_BTREE *cbt, WT_ITEM *srch_key, bool insert, WT_REF *le
      * In some cases we expect we're comparing more than a few keys with matching prefixes, so it's
      * faster to avoid the memory fetches by skipping over those prefixes. That's done by tracking
      * the length of the prefix match for the lowest and highest keys we compare as we descend the
-     * tree. The high boundary is reset on each new page, the lower boundary is maintained.
```
Comment (Siddhartha8899) [inline src/btree/row_srch.c:392]: Perhaps we could modify the comment as 
`The high boundary is reset on each new page, and the lower boundary is maintained unless the page index changes before reaching the leaf page.`

File: src/include/btree_inline.h (hunk 1 above)
Comment (Siddhartha8899) [reply src/include/btree_inline.h:2161]: However, seem like this function is also called at the internal page level. Wouldn't the assertion be significant there?

Comment (quchenhao) [reply src/include/btree_inline.h:2161]: We are just trying to get the new page index. We don't care if it is locked for split or not. So I think it should be ok.

File: src/btree/row_srch.c (hunk 2 above)
Comment (quchenhao) [reply src/btree/row_srch.c:392]: I also modify only to reset the high boundary if there is a race

**Comment by quchenhao** [comment]
mongodb patch build https://spruce.mongodb.com/version/645d91ec7742ae3793cad38b/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

**Comment by sulabhM** [comment]
I have suggested a new ticket and PR title, please edit if not convinced.

**Comment by quchenhao** [comment]
mongodb perf test https://spruce.mongodb.com/version/645d923f0ae606ff137b7fe9/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

**Review by agorrod — COMMENTED** [review]
Thanks @quchenhao - I've got a few questions about the changes you made.

I'm also curious: Are there other places in WiredTiger that do similar things? I see you reviewed column store implementation, which is the obvious one, but figured it was worth asking.

Is there a test case/change that can be added to make the fix more obvious?

File: src/btree/row_srch.c [hunk 3]
```diff
@@ -651,11 +654,15 @@ __wt_row_search(WT_CURSOR_BTREE *cbt, WT_ITEM *srch_key, bool insert, WT_REF *le
...
-         * parent's search to be wrong for the child. We only need to reset the high count, the page
-         * can only be extended so the low count is still correct.
+         * on the child). However, we can insert smaller (larger) keys to leaf pages at the start
+         * (end) of the tree, causing the parent's search to be wrong for the child. In this case,
+         * we need to reset the boundary.
+         *
+         * It is not easy to detect which case we are exactly in with the available information.
+         * Simply reset both the low and high boundary if we detect a change of tree structure.
          */
-        skiphigh = 0;
```
Comment (ajmorton) [inline src/btree/row_srch.c:658]: Is it ok to move `skiphigh = 0` inside the if branch here? Looking at [our description of WT-6257](https://jira.mongodb.org/browse/WT-10961?focusedCommentId=5416622&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-5416622) it seems like we could hit WT-6257 without the pindex changing as we only need to delete the rightmost page and insert one more key to hit the issue.

The same question applies to the internal page logic on line 506 if this is accurate.

File: src/btree/row_srch.c [hunk 4]
```diff
@@ -495,13 +496,15 @@ __wt_row_search(WT_CURSOR_BTREE *cbt, WT_ITEM *srch_key, bool insert, WT_REF *le
...
-             * the low count is still correct.
+             * least N bytes on the child). However, if a child internal page was split up or
+             * reverse split into the parent, the child page's key space will have been changed, and
+             * the values from the parent's search may be wrong for the child.
+             *
+             * It is not easy to detect which case we are exactly in with the available information.
+             * Simply reset both the low and high boundary if we detect a change of tree structure.
              */
-            skiphigh = 0;
+            if (parent_pindex != NULL && __wt_split_descent_race(current, parent_pindex))
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/btree/row_srch.c:506]: It's when we are at the root.

File: src/btree/row_srch.c [hunk 5]
```diff
@@ -389,9 +390,9 @@ __wt_row_search(WT_CURSOR_BTREE *cbt, WT_ITEM *srch_key, bool insert, WT_REF *le
      * In some cases we expect we're comparing more than a few keys with matching prefixes, so it's
      * faster to avoid the memory fetches by skipping over those prefixes. That's done by tracking
      * the length of the prefix match for the lowest and highest keys we compare as we descend the
-     * tree. The high boundary is reset on each new page, the lower boundary is maintained.
+     * tree. The boundaries are reset if the parent's page index changes concurrently.
      */
-    skiplow = 0;
+    skiplow = skiphigh = 0;
```
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/btree/row_srch.c:395]: We need to initialize skiphigh otherwise the compiler will complain.

File: src/include/btree_inline.h (hunk 1 above)
Comment (quchenhao) [reply src/include/btree_inline.h:2161]: We check the race when we have descended to the child page. At this stage, we have lost our hazard pointer to the internal page. Therefore, it is no longer "SAFE".

**Comment by quchenhao** [comment]
> Thanks @quchenhao - I've got a few questions about the changes you made.
> 
> I'm also curious: Are there other places in WiredTiger that do similar things? I see you reviewed column store implementation, which is the obvious one, but figured it was worth asking.
> 
> Is there a test case/change that can be added to make the fix more obvious?


I've done a global search of sikplow and only saw prefix optimization is used in insert list and row search. Therefore, I don't think there are other places that may encounter the same problem.

It is a complex race. I cannot think of a case that we can reliably reproduce this in python. I think it is possible to create a csuite test for this but that takes quite a bit of effort. May be we can create a separate ticket to do that. I've added the failure config to the pr so for now at least it is covered in some testing.

File: src/include/btree_inline.h (hunk 1 above)
Comment (agorrod) [reply src/include/btree_inline.h:2161]: I think the naming causes confusion here. It is only OK to call `WT_INTL_INDEX_GET_SAFE` when you know the memory couldn't possibly be freed by anything else (for example, if you are doing a split and know nothing else can be, or have exclusive access to the tree).
All other users need to call the non-safe version - if they don't then a split can cause unexpected behavior.

Regardless - I'm pretty sure that all callers of split_descent_race should meet the requirements. If they had exclusive access, they shouldn't need to check for structural changes.

Comment (quchenhao) [reply src/include/btree_inline.h:2161]: yes, I agree.

File: src/btree/row_srch.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (sueloverso) [reply src/btree/row_srch.c:395]: Minor nit: assuming this change stays, please alphabetize the variables on the line.

File: src/btree/row_srch.c [hunk 6]
```diff
@@ -428,7 +429,7 @@ __wt_row_search(WT_CURSOR_BTREE *cbt, WT_ITEM *srch_key, bool insert, WT_REF *le
          * Discard the currently held page and restart the search from the root.
          */
         WT_RET(__wt_page_release(session, current, 0));
-        skiplow = 0;
+        skiplow = skiphigh = 0;
```
Comment (sueloverso) [inline src/btree/row_srch.c:432]: Same nit here. Please alphabetize on the line.

File: src/btree/row_srch.c [hunk 7]
```diff
@@ -495,13 +496,15 @@ __wt_row_search(WT_CURSOR_BTREE *cbt, WT_ITEM *srch_key, bool insert, WT_REF *le
...
+             * least N bytes on the child). However, if a child internal page was split up or
+             * reverse split into the parent, the child page's key space will have been changed, and
+             * the values from the parent's search may be wrong for the child.
+             *
+             * It is not easy to detect which case we are exactly in with the available information.
+             * Simply reset both the low and high boundary if we detect a change of tree structure.
              */
-            skiphigh = 0;
+            if (parent_pindex != NULL && __wt_split_descent_race(current, parent_pindex))
+                skiplow = skiphigh = 0;
```
Comment (sueloverso) [inline src/btree/row_srch.c:507]: Alphabetize here.

File: src/btree/row_srch.c [hunk 8]
```diff
@@ -651,11 +654,15 @@ __wt_row_search(WT_CURSOR_BTREE *cbt, WT_ITEM *srch_key, bool insert, WT_REF *le
...
+         * on the child). However, we can insert smaller (larger) keys to leaf pages at the start
+         * (end) of the tree, causing the parent's search to be wrong for the child. In this case,
+         * we need to reset the boundary.
+         *
+         * It is not easy to detect which case we are exactly in with the available information.
+         * Simply reset both the low and high boundary if we detect a change of tree structure.
          */
-        skiphigh = 0;
+        if (__wt_split_descent_race(cbt->ref, pindex))
+            skiplow = skiphigh = 0;
```
Comment (sueloverso) [inline src/btree/row_srch.c:665]: and here

**Review by Siddhartha8899 — APPROVED** [review]

**Review by Sean04 — APPROVED** [review]

File: src/btree/row_srch.c [hunk 9]
```diff
@@ -492,16 +483,11 @@ __wt_row_search(WT_CURSOR_BTREE *cbt, WT_ITEM *srch_key, bool insert, WT_REF *le
...
-             * parent, the child page's key space will have been truncated, and the values from the
-             * parent's search may be wrong for the child. We only need to reset the high count
-             * because the split-page algorithm truncates the end of the internal page's key space,
-             * the low count is still correct.
+             * In some cases we expect we're comparing more than a few keys with matching prefixes,
+             * so it's faster to avoid the memory fetches by skipping over those prefixes. That's
+             * done by tracking the length of the prefix match for the lowest and highest keys.
              */
-            skiphigh = 0;
+            skiphigh = skiplow = 0;
```
Comment (ajmorton) [inline src/btree/row_srch.c:490]: Can you mention in this comment that `skiphigh` and `skiplow` are always set to 0 when we look at a page? Removing the default initialization of `skiplow` up above was throwing me off.
Same comment where we set these variables on line 641

Comment (agorrod) [reply src/btree/row_srch.c:490]: It's worth keeping context in the comment regarding why they must be reset across page traversals. I'd hate to add a similar optimization again in the future if we forgot.

I also asked a question in the Jira ticket about this solution to make sure I fully understand the issue.

File: src/btree/row_srch.c [hunk 10]
```diff
@@ -648,14 +641,18 @@ __wt_row_search(WT_CURSOR_BTREE *cbt, WT_ITEM *srch_key, bool insert, WT_REF *le
...
+         * tracking the length of the prefix match for the lowest and highest keys.
+         *
          * Reset the skipped prefix counts; we'd normally expect the parent's skipped prefix values
          * to be larger than the child's values and so we'd only increase them as we walk down the
          * tree (in other words, if we can skip N bytes on the parent, we can skip at least N bytes
-         * on the child). However, leaf pages at the end of the tree can be extended, causing the
-         * parent's search to be wrong for the child. We only need to reset the high count, the page
-         * can only be extended so the low count is still correct.
+         * on the child). However, if a child internal page was split up into the parent, the child
+         * page's key space will have been truncated, and the values from the parent's search may be
```
Comment (quchenhao) [inline src/btree/row_srch.c:652]: I think we should use ancestor instead of just parent, which is more accurate.

**Comment by sulabhM** [comment]
I have removed myself as a reviewer, I am helping with the processes for this ticket and observing changes but won't be required for a review.

**Review by agorrod — APPROVED** [review]
lgtm

File: src/btree/row_srch.c (hunk 10 above)
Comment (Siddhartha8899) [reply src/btree/row_srch.c]: Updated in [5642f0d](https://github.com/wiredtiger/wiredtiger/pull/9152/commits/5642f0d63871cc24f22d47a48b3d127e6cf0e715)

File: src/btree/row_srch.c (hunk 9 above)
Comment (ajmorton) [reply src/btree/row_srch.c:501]: @agorrod @quchenhao Sid and I have discussed this comment and are happy with the current iteration.

@quchenhao we've dropped "ancestor page" in favour of using parent and its child which I think conveys the same information about the relation to grandparent pages, but feel free to request ancestor again if you think it's clearer.
@agorrod we've changed the wording about why we need to reset skiphigh/skiplow but I think kept the intention behind the comment. Again feel free to request the original comment if it's clearer.

**Review by ajmorton — APPROVED** [review]
LGTM

**Pull request merged**